doc: add an howto about rule generation (#10154)
ocaml/dune: doc/howto/rule-generation.rst

rules are included in the main ``dune`` file, so ``dune runtest`` will run the
tests. Finally, the generated file is part of the source repository, so it is
not necessary to run several commands to run the test suite.

Let's expand a bit on how to achieve this.

Generating a ``dune.inc`` File
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Create a ``gen`` subdirectory and create a ``gen.ml`` file in it:

``gen/gen.ml``
  .. code:: ocaml

    let generate_rules base =
      Printf.printf
        {|
            (rule
             (with-stdout-to %s.gen
              (run %%{bin:tool} %s.input)))
        
            (rule
             (alias runtest)
             (action
              (diff %s.output %s.gen)))
             |}
        base base base base

    let () =
      Sys.readdir "."
      |> Array.to_list
      |> List.sort String.compare
      |> List.filter_map (Filename.chop_suffix_opt ~suffix:".input")
      |> List.iter generate_rules

Create a ``dune`` file in that directory:

``gen/dune``
  .. code:: dune

    (executable
     (name gen))

This defines an executable that lists ``*.input`` files in the current
directory and outputs rules on its standard output.

.. note::

>    It is important to sort the input files to ensure that the output is
>    independent on the order the files are returned by ``Sys.readdir``.

Reviewer: ```suggestion
   It is important to sort the input files to ensure that the output is
   independent from the order ``Sys.readdir`` returns the files .
```

Author: Don't we need to say
> from the order *in which* `Sys.readdir` returns the files?

(the context is that `Sys.readdir` returns a list of files, but in a seemingly random order)

Reviewer: Good call! I updated the suggestion.
